### PR TITLE
Added String Search MircroByte

### LINKED ISSUE #9: String Search
**MicroByte Title:** String Search

**Domain:** DSA

**Scope of learning for learners:** Understanding the working of KMP and Rabin Karp Algorithm

**Activities:** Compare performance each algorithm: Best/Average/Worst

### PR DESCRIPTION
**Pull request by IamShubhamGupto**
Closes #9 
Checklist:
- [x] Fixed file structure
- [x] Changed theory from paragraphs to points
- [x] Changed content from rather giving everything to questions and facts. Present in between actvities
- [x] Reduced content of actvity 1
- [x] Added gif visualization of KMP and Rabin Karp

Revision 3
Checklist
- [x] Made points in introduction to algorithms shorter.
- [x] Added Best,Average,Worst case complexity notations as Facts under each activity.
- [x] By default this is **disabled** but users can now see the text and pattern being matched. Since the text and pattern are really long texts, they appear wrapped in the terminal. change DEBUG macro to 1
- [x] Change 4 was the same as Change 2 from what i understood, hence marked as completed. 

Let me know if i missed something 🕵️‍♂️

### CODE REVIEW

**Comment by amoghrajesh** [comment]
Hey @IamShubhamGupto, you can make those template changes that I requested in the linked issue here itself by clicking on edit option. 
It helps us in managing the incoming PRs. 
Try to do this till we review your work 😄

**Review by amoghrajesh — CHANGES_REQUESTED** [review]
Hey @IamShubhamGupto!
I was reviewing your content.
You have posted pretty amazing work here. I am happy as a reader.
However, I would like to request a few changes to your work to get it into a more Learn By Doing method.

1. Your file structure is little bit off, all the images are in a folder but the folder is names as 'resources'. Kindly rename that to 'images'. The code files should be kept in another folder called 'src'.
2. You can cut down the initial amount of theory (explaining like the history of these algos) you have provided and rather provide just key points.
3. It is impressive that you are covering the asymptotic complexities by using time librarries and all. What you can add to this is try to map it to the usual complexities in terms of big-O notation or leave these as questions. (I would prefer the latter, but upto you)
4. In Activity 1 too, the theory can be simplified into crisp points so that learners don't get overwhelmed. 
5. Try to include some image/video references if possible 😄 

Really impressive work apart from these small changes.
Kudos man 😉

**Comment by IamShubhamGupto** [comment]
> Hey @IamShubhamGupto!
> I was reviewing your content.
> You have posted pretty amazing work here. I am happy as a reader.
> However, I would like to request a few changes to your work to get it into a more Learn By Doing method.
> 
> 1. Your file structure is little bit off, all the images are in a folder but the folder is names as 'resources'. Kindly rename that to 'images'. The code files should be kept in another folder called 'src'.
> 2. You can cut down the initial amount of theory (explaining like the history of these algos) you have provided and rather provide just key points.
> 3. It is impressive that you are covering the asymptotic complexities by using time librarries and all. What you can add to this is try to map it to the usual complexities in terms of big-O notation or leave these as questions. (I would prefer the latter, but upto you)
> 4. In Activity 1 too, the theory can be simplified into crisp points so that learners don't get overwhelmed.
> 5. Try to include some image/video references if possible 😄
> 
> Really impressive work apart from these small changes.
> Kudos man 😉

Heyy thanks for the review !! 😄 
Sure will make changes 1,2,4,5
in point 3, i didnt quite get what to change. Like each activity says which notation we will be comparing 😛

**Comment by IamShubhamGupto** [comment]
> Hey @IamShubhamGupto, you can make those template changes that I requested in the linked issue here itself by clicking on edit option.
> It helps us in managing the incoming PRs.
> Try to do this till we review your work 😄

Yes, before the next review it will be done

**Comment by amoghrajesh** [comment]
Hey @IamShubhamGupto talking about the 3, I meant at the end of every activity to compare the best case/worst case - you can mention at the end that the complexity of Rabin Karp is O(n+m) or something like that for best case and maybe the worst case is soemthing like O(something else) for both the algos 😄

**Comment by Ak-Shaw** [comment]
@IamShubhamGupto Kindly convert this PR to a draft and convert it back when you want us to review your work.

**Comment by IamShubhamGupto** [comment]
> @IamShubhamGupto Kindly convert this PR to a draft and convert it back when you want us to review your work.

Hey @Ak-Shaw I have made the changes requested. You guys can re review the PR now

File: string_search/README.md
Comment (amoghrajesh) [inline string_search/README.md:28]: Can you cut down the length of sentences by a bit - both in this algo and next one too.
Doing this will make it quite crisp and easy to read, you can drop the history of it as well.

**Comment by IamShubhamGupto** [comment]
> @IamShubhamGupto Kindly convert this PR to a draft and convert it back when you want us to review your work.

Hey @Ak-Shaw, I have made the requested changes, You guys can re review the PR

**Review by amoghrajesh — CHANGES_REQUESTED** [review]
Hey @IamShubhamGupto you have given a good start to the requests made earlier.
I still have some smaller feedbacks:

1. I have pointed out some changes in the readme regarding the length of sentences.
2. After the best, worst and average case it would be great if you could add the time complexities here
3. I see that a pair of strings are generated for each case - but I am not sure which ones being used, can you print those out too in the terminal if it doesn't require major changes to your implementation.
4. Last one - Will you be able to raise learner issues on this MicroByte once it gets merged. (Keep some of these in mind)
(For eg: You can have a question so that learners can verify your algo using two dummy strings or and let's say we say that the algo is O(m+n) then learners have to prove it by taking lengths of pattern and text and then dry run it and verify it)

Good work otherwise, I really appreciate the work 😃

File: string_search/README.md
Comment (IamShubhamGupto) [reply string_search/README.md:28]: Will do, The authors of the algorithm were added as people generally dont know their names.
Changes will be added in the next iteration

**Comment by IamShubhamGupto** [comment]
@amoghrajesh Regarding 3, The reason why the text and pattern were not included in the terminal was that text is a 1000 character string and pattern is a 100 character string. It is very messy with the line wraps.
With the current terminal implementation, the user sees only the time taken to execute the algorithm. Let me know if i should still do it.
Regarding 4, As much as i understood, given 2 dummy string, compute the time complexity value and not just in notation. I havent tried this before. Will look into it if i can implement

**Comment by amoghrajesh** [comment]
@IamShubhamGupto you can drop 3 if it gets too hard.
About 4, I meant that you need to mention in notation terms what the particular algo is running in - like maybe mention that KMP runs in O(n) in average case. Like this, maybe put it up for best, average and worst 👍🏽

**Review by amoghrajesh — APPROVED** [review]
Hey @IamShubhamGupto!
This is pretty good content now.
Thanks for making the time to time changes as requested by me.
I really like the structure of this particular MicroByte right now. (Especially with the time complexity notations for each of the cases)
I approve of this one, let's see what the other reviewers think.

You can start thinking of some learner issues that can be raised on this particular MicroByte once it gets merged.
Again, good work man 👍🏽

**Comment by IamShubhamGupto** [comment]
> Hey @IamShubhamGupto!
> This is pretty good content now.
> Thanks for making the time to time changes as requested by me.
> I really like the structure of this particular MicroByte right now. (Especially with the time complexity notations for each of the cases)
> I approve of this one, let's see what the other reviewers think.
> 
> You can start thinking of some learner issues that can be raised on this particular MicroByte once it gets merged.
> Again, good work man 👍🏽

Yayy, thanks !!
Yeah sure. is there any examples for learner issues i can look at first ?

**Comment by amoghrajesh** [comment]
Hey @IamShubhamGupto for an example of a learner issue, have a look at this one - https://github.com/Crio-Bytes/Miscellaneous/issues/60
You can also go to repos of your choice and under the issues tab, you will find many such issues 😄

**Review by Ak-Shaw — APPROVED** [review]
@IamShubhamGupto Interesting content you've come up with. I'm sure it'll be really beneficial for the learners to checkout the run time of the implemented algorithm through your code. 

Looking forward to more awesome contributions! :rocket:

**Comment by Ak-Shaw** [comment]
@all-contributors kindly add @IamShubhamGupto for:

- [x] code
- [x] content
- [x] ideas
- [x] doc

**Comment by IamShubhamGupto** [comment]
@Ak-Shaw Thank you !! This was one of the biggest projects ive worked on on github. Even I do look forward to adding more content

**Comment by Ak-Shaw** [comment]
@IamShubhamGupto Kindly showcase your new MicroByte in our slack workspace using [this](https://github.com/Crio-Bytes/Demo-Repo/blob/master/Maintainer%20Resources/new-microbyte-showcase-template.md) template.

**Pull request merged**
